### PR TITLE
[WIP] discover public URL

### PR DESCRIPTION
**Pull request by deads2k**
@simo5 this is where my head is at.  I have more wiring to finish, but the console public url is fully discoverable.

### CODE REVIEW

File: pkg/cmd/server/origin/webconsole_proxy.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+	return newValue
+}
+
+func (a *webConsolePublicURLAccessor) run(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		case <-time.After(10 * time.Second):
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/cmd/server/origin/webconsole_proxy.go]: That value can change over time

File: pkg/cmd/server/origin/webconsole_proxy.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+func (a *webConsolePublicURLAccessor) getPublicConsoleURL() string {
+	currValue, ok := a.publicURL.Load().(string)
+	if ok && len(currValue) > 0 {
+		return currValue
+	}
+
+	// if we aren't already set, try a live update
+	return a.updatePublicConsoleURL()
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/cmd/server/origin/webconsole_proxy.go]: The atomic value makes sure we only one writer wins and having a stale value is ok

File: pkg/cmd/server/origin/webconsole_proxy.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		if len(webconsolePublicURL) == 0 {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		http.Redirect(w, req, webconsolePublicURL, http.StatusFound)
+	})
+}
+
+func (c *MasterConfig) withConsoleRedirection(handler, assetServerHandler http.Handler, webconsolePublicURL string) http.Handler {
```
Comment (enj) [inline pkg/cmd/server/origin/webconsole_proxy.go]: Dead code now or just WIP?

File: pkg/cmd/server/origin/webconsole_proxy.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		return currValue
+	}
+
+	// if we aren't already set, try a live update
+	return a.updatePublicConsoleURL()
+}
+
+func (a *webConsolePublicURLAccessor) updatePublicConsoleURL() string {
+	// check to see if someone already did our work
+	currValue, ok := a.publicURL.Load().(string)
```
Comment (enj) [inline pkg/cmd/server/origin/webconsole_proxy.go]: Does not this prevent you from ever getting an updated value?

**Comment by simo5** [comment]
See #19194 for further progress

**Pull request closed without merging**